### ISSUE TITLE
As of v0.22.0, `no-shadow` errors on function names

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#2645]
The `no-shadow` rule previously allowed:
```js
var a = function a() {};
```

However, this appears to now warn, since the function's lexical name binding is technically shadowing the outer variable of the same name.

This is new breaking behavior - could there be a `no-shadow` option that ignores function names?

**Comment by michaelficarra** [comment]
See https://github.com/eslint/eslint/issues/2556. This was considered a bug fix, not a breaking change. Why are you writing code like that?

**Comment by ljharb** [comment]
To both name a function and avoid the hoisting caused by a declaration. I've not only always written my functions this way, but I've long advocated others do it too.

**Comment by ljharb** [comment]
I certainly could name the function differently than its containing variable, but this should be configurable so I'm not forced to.

**Comment by nzakas** [comment]
@ljharb this is trickier than it seems - what is the pattern that should be excluded here? We can't just say function expression names don't count, as that will have a bunch of side effects (this was the bug @michaelficarra is referring to).

What about this?

```
var a;
a = function a() {};
```

(Note: semantically equivalent to your example.)

**Comment by ljharb** [comment]
Essentially yes, I want an option so that function expression names don't count. Your example would be allowed in this case as well.

I would want to know that function declaration names are shadowing, since that could be accidental, but if I'm assigning a named function to a same-named variable, that's explicit and intentional, and I don't want that to set off a warning.

**Comment by mysticatea** [comment]
@ljharb I guess your example code wasn't warned by `no-shadow` rule. This rule ignores this general pattern.

```
> eslint -v
v0.22.1
> echo "var a = function a() {};" | eslint --stdin --no-eslintrc --reset --rule "{no-shadow:2}"
```

No errors as same as you expected.
But now, if the declaration was separated, it's warned.

```
> echo "var a; a = function a() {};" | eslint --stdin --no-eslintrc --reset --rule "{no-shadow:2}"

<text>
  1:21  error  a is already declared in the upper scope  no-shadow
```

Please see also #2608 and [a test for this pattern](https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-shadow.js#L24)

@nzakas If it's need, I can work to allow shadowing on `AssignmentPattern`/`AssignmentExpression`. But probably it's complex...

**Comment by ljharb** [comment]
@mysticatea aha, thanks for clearing that up. This also happens with the following:
var Foo = (function () {
   /* some setup */
  return function Foo() { };
}());
```

which is a pattern I use in the es* shims frequently.

**Comment by mysticatea** [comment]
@ljharb Oh, wow, I hadn't thought that pattern...

**Comment by mysticatea** [comment]
FYI, probably required fixes to allow the pattern are about [this scopes checking](https://github.com/eslint/eslint/blob/master/lib/rules/no-shadow.js#L53).

**Comment by nzakas** [comment]
@ljharb so you literally want all function expression names exempted from this check regardless of where they appear?

**Comment by ljharb** [comment]
Yes, I'd certainly be content with an option that exempted all function expression names from this check.

If there's a practical way to have more granular configurations, I'm always all for that :-)

**Comment by nzakas** [comment]
In this case it's really all or nothing. I'm okay with adding suffering to exempt all function each expression names, it just means you'll miss things like this:

```js
var b = function a() {
    var a;     // no warning
};
```

**Comment by ljharb** [comment]
As a configurable option that'd be great :-)

**Comment by mysticatea** [comment]
I think this issue is similar to the TDZ issue.
To warn shadowing on the uninitialized variable is problem, so I don't think to add the option is good.
Would ignoring shadowing of the declaring variable in the `VariableDeclarator` be not enough?

**Comment by ljharb** [comment]
The TDZ doesn't apply to function name bindings, only to "let" and "const", so I'm not sure how this relates?

**Comment by mysticatea** [comment]
I thought this is an issue about "uninitialized variables and shadowing", so felt similar.

**Comment by nzakas** [comment]
Lol, I see my autocorrect put "suffering" in my last reply - how Freudian. Okay, let's add an option (off by default) to exclude function expression names from the rule. Suggestions on the option name?

**Comment by ljharb** [comment]
Maybe `"functionExpressions": false` defaults to `true`, or `"excludeFunctionExpressions": true` defaults to `false`?

**Comment by nzakas** [comment]
Naming is hard, because it's not about function expressions, it's about their names. So maybe `excludeFunctionExpressionNames`? (Too long?)

**Comment by ljharb** [comment]
I like `excludeFunctionExpressionNames` - letters are free, and config keys typed once, so I'm not concerned about length :-)

**Comment by michaelficarra** [comment]
This exception seems so arbitrary to me. Are we going to add an exception for every declaration form? The justification for this exception is weak in my opinion.

**Comment by ljharb** [comment]
@michaelficarra what is the concern caused by shadowing in the first place, as you see it?

**Comment by michaelficarra** [comment]
I'm not arguing for/against this rule in general. Shadowing is fine in my book. I'm just failing to see why someone would want to enable this rule but also follow a pattern that necessitates shadowing.

**Comment by ljharb** [comment]
My rationale is that variable shadowing is often accidental, but function name shadowing is often intentional and explicit. Others' mileage may vary, which is why I'm happy to have it be configurable.

**Comment by nzakas** [comment]
@michaelficarra we always handle exceptions on a case-by-case basis.

**Comment by ljharb** [comment]
One other thing to note:
```js
var a = 3;
function a() {}
```

This will error as-is, because the function name is technically shadowing the variable. However, nothing inside the function references "a". Even in the default case, is there any value on warning here?

**Comment by mysticatea** [comment]
@ljharb I think it is re-declaration. It's reported by [no-redeclare](http://eslint.org/docs/rules/no-redeclare).

**Comment by ljharb** [comment]
Ah, perhaps in that case. How about:
```js
var a = 3;
(function () {
  function a() {}
}());
```

**Comment by nzakas** [comment]
@ljharb please include the actual  ESLint output. "This will error" is not enough information.

**Comment by ljharb** [comment]
Sure - the above code, with a default configuration, outputs:
```bash
temp.js
  1:4   error  a is defined but never used               no-unused-vars
  2:1   error  Missing "use strict" statement            strict
  3:11  error  a is already declared in the upper scope  no-shadow
  3:11  error  a is defined but never used               no-unused-vars

✖ 4 problems (4 errors, 0 warnings)
```

The third error is the one I'd like to avoid without disabling `no-shadow` completely about.

**Comment by nzakas** [comment]
Got it, thanks

**Comment by evilebottnawi** [comment]
@nzakas 
```javascirpt
var Environment = (function () {
    function Environment() {
        // Some logins
    }
});

module.exports = Environment;
```
Add also error `no-shadow`, in my opinion it also referred to the problem

**Comment by ljharb** [comment]
This is still a problem - `echo 'var foo; foo = function foo() {};' | eslint --stdin --no-eslintrc --rule 'no-shadow: 2'` on v2.2 also prints out an error - i'm explicitly assigning a function expression to the variable it's shadowing, and I'd expect this to be allowed.

**Comment by nzakas** [comment]
@ljharb the issue is still open, so that is expected. Want to submit a PR?

**Comment by nzakas** [comment]
Unfortunately, it looks like no one was willing to work on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that "help wanted" issues that aren't addressed within 90 days tend to never get addressed, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to and not something that members of community were willing to implement.
